src/coreclr/src/nativeaot/System.Private.Reflection.Execution/src/Internal/Reflection/Execution/FieldAccessors/StaticFieldAccessor.cs

```diff
@@ -51,6 +56,11 @@ public sealed override void SetFieldDirect(TypedReference typedReference, object
             {
                 RuntimeAugments.EnsureClassConstructorRun(_cctorContext);
             }
+            if (IsFieldInitOnly)
```

Comment:
Could you move this check to `WritableStaticFieldAccessor.SetFieldBypassCctor` (same for the Direct version.

We want to make sure that:

```csharp
typeof(Foo).GetField("F").SetValue(null, "hello");

struct Foo
{
    public static readonly int F = 12;
}
```

Will keep throwing the "System.ArgumentException: 'Object of type 'System.String' cannot be converted to type 'System.Int32'.'" exception (i.e. check the value being set first).